### PR TITLE
[SystemZ][z/OS] __ptr32 support for z/OS in Clang

### PR DESCRIPTION
**Pull request by fanbo-meng**
Enabling __ptr32 keyword to support in Clang for z/OS. It is represented by addrspace(1) in LLVM IR. Unlike existing implementation, __ptr32 is not mangled into symbol names for z/OS.

### CODE REVIEW

**Comment by fanbo-meng** [comment]
code-formatter is complaining about the entire enum table. To be less pervasive, I'll not address that.

File: clang/lib/Basic/IdentifierTable.cpp [hunk 1]
```diff
@@ -107,12 +107,14 @@ namespace {
...
+    KEYZOS        = 0x2000000,
+    KEYNOZOS      = 0x4000000,
+    KEYHLSL       = 0x8000000,
+    KEYFIXEDPOINT = 0x10000000,
     KEYMAX        = KEYFIXEDPOINT, // The maximum key
     KEYALLCXX = KEYCXX | KEYCXX11 | KEYCXX20,
-    KEYALL = (KEYMAX | (KEYMAX-1)) & ~KEYNOMS18 &
-             ~KEYNOOPENCL // KEYNOMS18 and KEYNOOPENCL are used to exclude.
+    KEYALL = (KEYMAX | (KEYMAX-1)) & ~KEYNOMS18 & ~KEYNOOPENCL &
+             ~KEYNOZOS // KEYNOMS18, KEYNOOPENCL, KEYNOZOS are used to exclude.
```
Comment (zibi2) [inline clang/lib/Basic/IdentifierTable.cpp:117]:
```suggestion
             ~KEYNOZOS // KEYNOMS18, KEYNOOPENCL, KEYNOZOS are excluded.
```

File: clang/lib/Basic/Targets/SystemZ.h [hunk 2]
```diff
@@ -224,6 +252,19 @@ class LLVM_LIBRARY_VISIBILITY SystemZTargetInfo : public TargetInfo {
   std::pair<unsigned, unsigned> hardwareInterferenceSizes() const override {
     return std::make_pair(256, 256);
   }
+
+  uint64_t getPointerWidthV(LangAS AddrSpace) const override {
+    unsigned TargetAddrSpace = getTargetAddressSpace(AddrSpace);
+    return (TargetAddrSpace == ptr32 && getTriple().isOSzOS() &&
+            getTriple().isArch64Bit())
```
Comment (zibi2) [inline clang/lib/Basic/Targets/SystemZ.h:259]:
```suggestion
    return (getTriple().isOSzOS() && getTriple().isArch64Bit() &&
            getTargetAddressSpace(AddrSpace) == ptr32)
```

File: clang/lib/Frontend/CompilerInvocation.cpp [hunk 3]
```diff
@@ -3567,6 +3567,14 @@ void CompilerInvocationBase::GenerateLangArgs(const LangOptions &Opts,
       GenerateArg(Consumer, OPT_ftrigraphs);
   }
 
+  if (T.isOSzOS()) {
+    if (!Opts.ZOSExt)
+      GenerateArg(Consumer, OPT_fno_zos_extensions);
```
Comment (zibi2) [inline clang/lib/Frontend/CompilerInvocation.cpp:3572]: I assume the default is -fno-zos-extensions, can you check why we need to explicitly pass this option?

File: clang/lib/Sema/SemaType.cpp [hunk 4]
```diff
@@ -7097,10 +7098,14 @@ static bool handleMSPointerTypeQualifierAttr(TypeProcessingState &State,
...
+    if (Triple.isOSzOS()) {
       ASIdx = LangAS::ptr32_uptr;
-    else
-      ASIdx = LangAS::ptr32_sptr;
+    } else {
+      if (Attrs[attr::UPtr])
+        ASIdx = LangAS::ptr32_uptr;
+      else
+        ASIdx = LangAS::ptr32_sptr;
+    }
```
Comment (zibi2) [inline clang/lib/Sema/SemaType.cpp:7108]: I assume this block can be simplifed

File: llvm/lib/Target/SystemZ/SystemZTargetMachine.cpp [hunk 5]
```diff
@@ -59,6 +59,14 @@ static std::string computeDataLayout(const Triple &TT) {
   // Data mangling.
   Ret += DataLayout::getManglingComponent(TT);
 
+  // Special features for z/OS.
+  if (TT.isOSzOS()) {
+    if (TT.isArch64Bit()) {
```
Comment (zibi2) [inline llvm/lib/Target/SystemZ/SystemZTargetMachine.cpp:64]: You can collapse these 2 `if` statements into one.

File: clang/lib/Sema/SemaType.cpp [hunk 6]
```diff
@@ -7097,10 +7098,14 @@ static bool handleMSPointerTypeQualifierAttr(TypeProcessingState &State,
     else if (Attrs[attr::UPtr])
       ASIdx = LangAS::ptr32_uptr;
   } else if (PtrWidth == 64 && Attrs[attr::Ptr32]) {
-    if (Attrs[attr::UPtr])
+    if (Triple.isOSzOS()) {
```
Comment (perry-ca) [inline clang/lib/Sema/SemaType.cpp:7101]:
```suggestion
    if (Attrs[attr::UPtr] || Triple.isOSzOS())
      ASIdx = LangAS::ptr32_uptr;
    else
      ASIdx = LangAS::ptr32_sptr;
```
This is simpler.

File: clang/test/CodeGen/zos-mixed-ptr-sizes-malloc.c [hunk 7]
```diff
@@ -0,0 +1,84 @@
+// RUN: %clang_cc1 -triple s390x-ibm-zos -fzos-extensions -emit-llvm -O2 < %s | FileCheck %s --check-prefix=X64
```
Comment (perry-ca) [inline clang/test/CodeGen/zos-mixed-ptr-sizes-malloc.c:1]: Does this test need to be some complicated?

**Comment by fanbo-meng** [comment]
Transferring this to @abhina-sree

**Pull request closed without merging**